### PR TITLE
Update GH PR template to mention grammar doc

### PR DESCRIPTION
**Pull request by JasonGross**
I figure it's better to not leave it up to reviewers to remember to
remind people to do this all the time.  See also
https://github.com/coq/coq/pull/14174#discussion_r620522218

### CODE REVIEW

**Comment by Zimmi48** [comment]
This PR is probably a good idea.

We should also add a linter check that the grammar files are up-to-date in CI. Now that I'm about to merge the Dune PR, it should make calling the `doc_gram_rsts` target easier. However, I think doc_grammar should remain an optional step in the build process, as this is only relevant to packagers and should not affect the build process for users installing Coq or packaging it.

**Comment by jfehrle** [comment]
The plan was to have every build run `doc_grammar` in verification mode, checking that the files are current and failing if they're not.  Perhaps that's exactly what you mean by "a linter check".

**Comment by Zimmi48** [comment]
Yes.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ Fixes / closes #????
 - [ ] Added / updated test-suite
 <!-- If this is a feature pull request / breaks compatibility: -->
 <!-- (Otherwise, remove these lines.) -->
-- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
+- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified, and including reflecting any changes to the grammar in `common.edit_mlg` and running `make doc_gram_rsts`).
```
Comment (Zimmi48) [inline .github/PULL_REQUEST_TEMPLATE.md:18]: Not all the grammar changes will require a change to `common.edit_mlg` (fortunately!). So I propose to simplify the wording to:
```suggestion
- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified, and including reflecting any changes to the grammar by running `make doc_gram_rsts`).
```

Comment (JasonGross) [reply .github/PULL_REQUEST_TEMPLATE.md:18]: How are users supposed to know if such an edit is necessary?  I have a sight preference for mentioning editing the file, so that people who edit the grammar can at least tag the doc team if they think they might me missing something.

Comment (JasonGross) [reply .github/PULL_REQUEST_TEMPLATE.md:18]: (In my mind, the goal of this checklist is to prevent as much as possible cases where people forget to synchronize or update something without realizing they've forgotten.)

Comment (Zimmi48) [reply .github/PULL_REQUEST_TEMPLATE.md:18]: >How are users supposed to know if such an edit is necessary?

If an edit is required, the `doc_gram_rsts` target should fail with an error message mentioning the file. I don't expect that there would be any case when the file should be edited but the target doesn't fail.

Comment (jfehrle) [inline .github/PULL_REQUEST_TEMPLATE.md:18]:
```suggestion
- [ ] If you modified Coq's grammar, run `make doc_gram_rsts` to update the syntax in the documentation (see `docgram/README.md`).
- [ ] Corresponding documentation was added / updated, including any changed warning and error messages.
```

I suggest splitting this--checklist items should kept short so they're easy to read.  I started with Théo suggested wording.

In practice, the doc reviewers will need to pay attention to the docgram changes.  It's not foolproof (fools generally being quite clever...).

**Review by jfehrle — COMMENTED** [review]
WDYT?

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (JasonGross) [reply .github/PULL_REQUEST_TEMPLATE.md:18]: This seems like magic; how does it work?  Does it even work with tactic notations that modify the grammar?

Comment (jfehrle) [reply .github/PULL_REQUEST_TEMPLATE.md]: The tool helps, but it doesn't do everything.  One should review the changes it makes.  The things it does:
- updated syntax in a `:tacn:` or `:cmd:` - does matching based on the tactic/command name.  This updating works well.
- adding or changing productions of an existing nonterminal shown in `.. insertprodn / .. prodn::` will appear automatically, but may need to be reordered for the sake of readability in `orderedGrammar`
- adding new nonterminals - almost always need to adjust `orderedGrammar` and/or `insertprodn` to show the nonterminal in a sensible place
- if an existing nonterminal is renamed, the ordering and text may need adjustment.  The tool is unable to identify renamings
- `:tacv:` and `:cmdv:` are not handled, but we're in the process of removing them

As we've been revising the doc, we've frequently edited productions to make them more readable.  Frequently we rename nonterminals (and later feed these changes back into the grammar), splice nonterminals (substitute the RHS in place of the LHS everywhere and drop the symbol).  We also edit parts of the grammar that are hard to understand--they may be written in a particular way so the parser can handle them or for developer convenience.

The tool only looks at `mlg` files.  It doesn't examine anything in `.v` files.  Of course, the set of available tactic notations depends on what files you have loaded.  Any Ltac tactics that we include in the doc are added in manually in `common.edit_mlg`.

Comment (jfehrle) [reply .github/PULL_REQUEST_TEMPLATE.md]: I'm sure it's possible to recover tactic notations from the parser (some assembly required).  But the doc currently doesn't cover notations, so not on my "to do" list.

**Review by jfehrle — APPROVED** [review]
LGTM.

@Zimmi48 any idea when you might get around to adding the linter check (in your copious free time)?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -15,7 +15,8 @@ Fixes / closes #????
 - [ ] Added / updated test-suite
 <!-- If this is a feature pull request / breaks compatibility: -->
 <!-- (Otherwise, remove these lines.) -->
-- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
+- [ ] If you modified Coq's grammar, run `make doc_gram_rsts` to update the syntax in the documentation (see `docgram/README.md`).
+- [ ] Corresponding documentation was added / updated, including any changed warning and error messages.
 - [ ] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).
```
Comment (Zimmi48) [inline .github/PULL_REQUEST_TEMPLATE.md:20]: We could try to be more consistent with the rest of the template (where most conditions are inside comments).
Also note that inserting this new checkbox just after "if this is a feature PR" doesn't really make sense. Let's have it with its own condition block.

```suggestion
- [ ] Corresponding documentation was added / updated, including any changed warning and error messages.
- [ ] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).
<!-- If you modified Coq's grammar (otherwise remove this line): -->
- [ ] Run `make doc_gram_rsts` to update the syntax in the documentation (see `docgram/README.md`).
```

**Review by Zimmi48 — COMMENTED** [review]
Sorry for nitpicking but I feel like nitpicking is important when we change the PR template that everybody will get.

**Comment by Zimmi48** [comment]
We might also want to change `make doc_gram_rsts` to `make -f Makefile.make doc_gram_rsts` to be more explicit for developers who have set `COQ_USE_DUNE=1`.

**Comment by Zimmi48** [comment]
@jfehrle We should probably stop the target from printing all the warnings by default, because they might be confusing for contributors (who might worry about having introduced some).

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (jfehrle) [reply .github/PULL_REQUEST_TEMPLATE.md:20]: OK, but I would put the doc_gram item before updating the doc--makes more sense to do the items in that order.  And maybe have updating the changelog first, since most but not all PRs change it.

OTOH, I wonder how many people read this for every PR they submit?  I've gotten to the point where deleting the template is a routine first step for my PRs.

**Comment by jfehrle** [comment]
There is a `-no-warn` command line option for `doc_grammar`.  The argument parsing is near the end of `doc_grammar.ml`.  Let me review the current messages, which I may not get to for a couple days.  And there may be small issues to deal with when we turn it on.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (Zimmi48) [reply .github/PULL_REQUEST_TEMPLATE.md:20]: > OTOH, I wonder how many people read this for every PR they submit? I've gotten to the point where deleting the template is a routine first step for my PRs.

Well, me too also I will sometimes keep the checkbox items that interest me. And the thing being there seems to be sufficient to actually make people do the requested steps (or at least it was, see Section 3.5.2 from [my thesis](https://www.irif.fr/_media/users/theo/memoirthesis.pdf)).

**Comment by Zimmi48** [comment]
> There is a `-no-warn` command line option for `doc_grammar`.

Sure, but that's not included in the make target by default.

**Comment by jfehrle** [comment]
I fixed `-no-warn` in #14262.  Is there anything else you'd like me to address?

**Comment by JasonGross** [comment]
I would like to cede control of this PR to those more well-informed.  Would one of you be willing to push whatever changes you want to see?

**Comment by Zimmi48** [comment]
Sounds good. I'll integrate the contents in my related PR.

**Comment by ppedrot** [comment]
Should this still go into 8.14?

**Comment by Zimmi48** [comment]
No. Actually, the contents will be moved into #14245 and I'm deliberately waiting that 8.14 is out of the way before finishing this PR.

**Comment by Zimmi48** [comment]
Now the PR template is updated as part of #14245.
